In [45]:
# Import necessary libraries
import pandas as pd
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import category_encoders as ce

# Load and clean the data
data = pd.read_csv("Dataset_Final_v6.csv")
data_cleaned = data.iloc[1:].copy()

# Encode features and column 'State' and split data
label_encoder = LabelEncoder()
data_cleaned['Country_encoded'] = label_encoder.fit_transform(data_cleaned['Country'])

data_cleaned['Type_encoded'] = label_encoder.fit_transform(data_cleaned['Type'])

data_cleaned['Sub-Category_encoded'] = label_encoder.fit_transform(data_cleaned['Sub-Category'])

data_cleaned['Restarts_encoded'] = label_encoder.fit_transform(data_cleaned['Restarts'])

data_cleaned['Cost overruns_encoded'] = label_encoder.fit_transform(data_cleaned['Cost overruns'])

data_cleaned['Time overruns_encoded'] = label_encoder.fit_transform(data_cleaned['Time overruns'])

data_cleaned['Content deficiency_encoded'] = label_encoder.fit_transform(data_cleaned['Content deficiency'])

data_cleaned['User Involvement_encoded'] = label_encoder.fit_transform(data_cleaned['User Involvement'])

data_cleaned['Executive Management_encoded'] = label_encoder.fit_transform(data_cleaned['Executive Management'])

data_cleaned['Statement of Requirements_encoded'] = label_encoder.fit_transform(data_cleaned['Statement of Requirements'])

data_cleaned['Planning_encoded'] = label_encoder.fit_transform(data_cleaned['Planning'])

data_cleaned['Expectations_encoded'] = label_encoder.fit_transform(data_cleaned['Expectations'])

data_cleaned['Requirements & Specifications_encoded'] = label_encoder.fit_transform(data_cleaned['Requirements & Specifications'])

# Initialize the encoder for ordinal data
encoder_2 = ce.OrdinalEncoder(cols=['Changing Requirements & Specifications'])

# Encode the 'Changing Requirements & Specifications' column
data_cleaned['Changing Requirements & Specifications_encoded'] = encoder_2.fit_transform(data_cleaned['Changing Requirements & Specifications'])

#data_cleaned['Changing Requirements & Specifications_encoded'] = label_encoder.fit_transform(data_cleaned['Changing Requirements & Specifications'])

data_cleaned['Technical Competence_encoded'] = label_encoder.fit_transform(data_cleaned['Technical Competence'])

data_cleaned['Resources Availability_encoded'] = label_encoder.fit_transform(data_cleaned['Resources Availability'])

data_cleaned['IT Management_encoded'] = label_encoder.fit_transform(data_cleaned['IT Management'])

data_cleaned['Technical Literacy_encoded'] = label_encoder.fit_transform(data_cleaned['Technical Literacy'])

data_cleaned['User satisfaction_encoded'] = label_encoder.fit_transform(data_cleaned['User satisfaction'])

label_encoder_state = LabelEncoder()
data_cleaned['State_encoded'] = label_encoder_state.fit_transform(data_cleaned['State'])

Start building the model with the first feature:

Predictor: 'User satisfaction'

In [46]:
X = data_cleaned[['User satisfaction_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 72.73%
Precision: 76.92%
Recall: 62.50%
F1 Score: 68.97%


Predictors: 'User satisfaction', 'Technical Literacy'

In [47]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 78.79%
Precision: 76.47%
Recall: 81.25%
F1 Score: 78.79%


Evaluation metrics went up, meaning the model's performance has been improved.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management'

In [48]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 84.85%
Precision: 78.95%
Recall: 93.75%
F1 Score: 85.71%


Evaluation metrics went up, meaning the model's performance has been improved.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Resources Availability'

In [49]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Resources Availability_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 81.82%
Precision: 75.00%
Recall: 93.75%
F1 Score: 83.33%


Evaluation metrics went down, so we should not use 'Resources Availability' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Technical Competence'

In [50]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Technical Competence_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 84.85%
Precision: 78.95%
Recall: 93.75%
F1 Score: 85.71%


Evaluation metrics remains unchanged, so we should not use 'Technical Competence' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications'

In [51]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 84.85%
Precision: 78.95%
Recall: 93.75%
F1 Score: 85.71%


Evaluation metrics went up, meaning the model's performance has been improved.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications'

In [52]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 83.33%
Recall: 93.75%
F1 Score: 88.24%


Evaluation metrics went up, meaning the model's performance has been improved.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Expectations'

In [53]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Expectations_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 83.33%
Recall: 93.75%
F1 Score: 88.24%


Evaluation metrics went down, so we should not use 'Expectations' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Planning'

In [54]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Planning_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 83.33%
Recall: 93.75%
F1 Score: 88.24%


Evaluation metrics went down, so we should not use 'Planning' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Statement of Requirements'

In [55]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Statement of Requirements_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 78.79%
Precision: 76.47%
Recall: 81.25%
F1 Score: 78.79%


Evaluation metrics went down, so we should not use 'Statement of Requirements' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Executive Management'

In [56]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Executive Management_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 80.00%
Recall: 100.00%
F1 Score: 88.89%


Evaluation metrics went down, so we should not use 'Executive Management' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'User Involvement'

In [57]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'User Involvement_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 80.00%
Recall: 100.00%
F1 Score: 88.89%


Evaluation metrics went down, so we should not use 'User Involvement' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency'

In [37]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 90.91%
Precision: 88.24%
Recall: 93.75%
F1 Score: 90.91%


Evaluation metrics went up, meaning the model's performance has been improved.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency', 'Time overruns'

In [38]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded',  'Time overruns_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 84.85%
Precision: 76.19%
Recall: 100.00%
F1 Score: 86.49%


Evaluation metrics went down, so we should not use 'Time overruns' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency', 'Cost overruns'

In [39]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded', 'Cost overruns_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 83.33%
Recall: 93.75%
F1 Score: 88.24%


Evaluation metrics remain unchanged, so we should not use 'Cost overruns' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency', 'Restarts'

In [40]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded', 'Restarts_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 84.85%
Precision: 82.35%
Recall: 87.50%
F1 Score: 84.85%


Evaluation metrics went down, so we should not use 'Restarts' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency', 'Duration (Years)'

In [41]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded', 'Duration (Years)']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 90.91%
Precision: 88.24%
Recall: 93.75%
F1 Score: 90.91%


Evaluation metrics went down, so we should not use 'Duration (Years)' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency', 'Sub-Category'

In [42]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded', 'Sub-Category_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 87.50%
Recall: 87.50%
F1 Score: 87.50%


Evaluation metrics went down, so we should not use 'Sub-Category' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency', 'Type'

In [43]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded', 'Type_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 90.91%
Precision: 84.21%
Recall: 100.00%
F1 Score: 91.43%


Evaluation metrics went down, so we should not use 'Type' as a predictor.

Predictors: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency', 'Country'

In [44]:
X = data_cleaned[['User satisfaction_encoded', 'Technical Literacy_encoded', 'IT Management_encoded', 'Changing Requirements & Specifications_encoded', 'Requirements & Specifications_encoded', 'Content deficiency_encoded', 'Country_encoded']]
y = data_cleaned['State_encoded']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the XGBoost classifier
clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
clf.fit(X_train, y_train)

# Predict on the testing set
y_pred = clf.predict(X_test)

# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label=1)
recall = recall_score(y_test, y_pred, pos_label=1)
f1 = f1_score(y_test, y_pred, pos_label=1)

# Print the evaluation metrics
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Precision: {precision * 100:.2f}%")
print(f"Recall: {recall * 100:.2f}%")
print(f"F1 Score: {f1 * 100:.2f}%")

Accuracy: 87.88%
Precision: 83.33%
Recall: 93.75%
F1 Score: 88.24%


Evaluation metrics went down, so we should not use 'Country' as a predictor.

In conclusion, the best model was built with these features: 'User satisfaction', 'Technical Literacy', 'IT Management', 'Changing Requirements & Specifications', 'Requirements & Specifications', 'Content deficiency'.

Evaluation metrics:

Accuracy: 90.91%

Precision: 88.24%

Recall: 93.75%

F1 Score: 90.91%